src/json-observe.ts

```diff
@@ -0,0 +1,262 @@
+                instance.defaultCallback({ op: 'replace', path: distPath, value: receiver });
+                return Reflect.set(target, key, receiver);
+              }
+              //when `undefined` property is set to something, should generate add (undefined in JSON.stringified to no value)
+              else //object
+              {
+                instance.defaultCallback({ op: 'add', path: distPath, value: receiver });
+                return Reflect.set(target, key, receiver);
+              }
+            }
+            else {
+              instance.defaultCallback({ op: 'replace', path: distPath, value: receiver });
+              return Reflect.set(target, key, receiver);
+            }
+          }
+          else
+            instance.defaultCallback({ op: 'add', path: distPath, value: receiver });
+
+          return Reflect.set(target, key, receiver);
+        },
+        deleteProperty: function (target, key) {
+
+          //when when an `undefined` property is deleted
+          if (typeof target[key] === 'undefined')
+            return Reflect.deleteProperty(target, key);
+
+          // else {
+          var distPath = path + '/' + instance.escapePathComponent(key.toString());
+          instance.defaultCallback({ op: 'remove', path: distPath });
+          return Reflect.deleteProperty(target, key);
+        }
+      });
+    return proxy;
+  }
+
+  //grab tree's leaves one by one, encapsulate them into a proxy and return
+  private _proxifyObjectTreeRecursively(root: any, path): any {
+    if(!path) path = "";
+    for (var key in root) {
+      if (root.hasOwnProperty(key)) {
+        var distPath = path + '/' + this.escapePathComponent(key);
+        if (typeof root[key] === 'object') {
+          root[key] = this.generateProxyAtPath(root[key], distPath);
+          this._proxifyObjectTreeRecursively(root[key], distPath);
+        }
+      }
+    }
+    return this.generateProxyAtPath(root, "/");
+  }
+
+  //this function is for aesthetic purposes
+  private proxifyObjectTree(root: any): any {
+    
+    /*
+    while proxyifying object tree,
+    the proxyifying operation itself is being
+    recorded, which in an unwanted behavior,
+    that's why we disable recording through this
+    inital process;
+    */
+
+    this.disableCallback = true;
+    var proxifiedObject = this._proxifyObjectTreeRecursively(root, "");
+
+    /* OK you can record now */
+    this.disableCallback = false;
+
+    return proxifiedObject;
+  }
+  private equalsEvents(op1, op2): boolean {
+    if (op1.value && op2.value)
+      return op1.op === op2.code && op1.path === op2.path && op1.value === op2.value;
+
+    return op1.op === op2.code && op1.path === op2.path;
+  }
+  constructor(root) {
+    this.originalObject = root;
+    this.cachedProxy = null;
+    this.patches = [];
+    this.isRecording = false;
+    this.userCallback;
+
+    this.defaultCallback = function (sender) {
+      return function (event) {
```

Comment:
Could you define `var sender = this` two lines above, to get rid of unnecessary closure?